eip-0014.md

```diff
@@ -0,0 +1,513 @@
+# Automated Decentralized Exchange
+
+* Author: kushti, Ilya Oskin
+* Status: Proposed
+* Created: 12-Mar-2021
+* Last edited: 16-Apr-2021
+* License: CC0
+* Track: Standards 
+
+## Motivation 
+
+Act of exchange without trusted parties is a most basic primitive for decentralized finance on top of blockchains. Thus contracts for that were introduced early, and basic single-chain swap contract was introduced early in the [ErgoScript whitepaper](https://ergoplatform.org/docs/ErgoScript.pdf). Then a lot of other order contracts appeared: with partial filling, buyback guarantee and so on. What is good for traders in decentralized worlds, such contracts are usually composable. 
+While swap order contracts allows for orderbook-based decentralized exchanges (DEXes), now popular AMM-based DEXes (where AMM stands for Automated Market Maker) are also possible on Ergo. 
+Interestingly, unlike other known blockchains, thanks to the extended UTXO model, liquidity pool contracts for AMM-based DEXes can be combined with order contracts (for orderbook-based DEXes). This gives unique possibility to have shared liquidity among different types of exchanges on top of the Ergo blockchain. 
+
+This PR provides a description of the Automated Decentralized Exchange protocol on top of the Ergo.
+
+## Order-book DEX
+
+Orders are waiting for another orders to be matched, or for a refund command. There're the following three types of orders — "buy" (i.e. buy tokens with ERG), "sell" (i.e. sell tokens for ERG), or "swap" (buy non-ERG tokens with other non-ERG tokens) orders. Order-book DEX has the advantage of working best for those pairs with high liquidity.
+
+### Atomic orders
+
+Atomic orders can only be executed completely. Such orders can be either be aggregated by the ErgoDEX client so that users can choose from them or matched with partial orders which will be defined next.
+
+#### Buy order [ERG -> Token]
+
+```scala
+{
+    val quoteId     = SELF.R4[Coll[Byte]].get // R4 - quote tokenId
+    val price       = SELF.R5[Long].get       // R5 - price per token
+    val feePerToken = SELF.R6[Long].get       // R6 - fee per token
+
+    val maybeRewardBox           = OUTPUTS(0)
+    val isValidRewardProposition = maybeRewardBox.propositionBytes == pk.propBytes
+    val maybeRewardToken         = maybeRewardBox.tokens(0)
+
+    val rewardTokens =
+        if (isValidRewardProposition && maybeRewardToken._1 == quoteId) maybeRewardToken._2
+        else 0L
+
+    val feeCharged    = rewardTokens * feePerToken
+    val isValidReward = (SELF.value.toBigInt - feeCharged) <= rewardTokens * price
```

Comment:
Shouldn't this be `>=` instead? As we need enough ERGs to pay for the tokens.

Reply:
This contract only ensures that user gets at least `SELF.value.toBigInt - feeCharged / price` rewardTokens.

Reply:
:+1: this one looks easy but took me a while, thanks